ansible_rulebook/websocket.py

```diff
@@ -58,68 +85,55 @@ async def _connect_websocket(
-            elif isinstance(e, OSError) and "[Errno 61]" in str(e):
-                # Sleep and retry implemention duplicated from
-                # websockets.lagacy.client.Client
-
-                # Add a random initial delay between 0 and 5 seconds.
-                # See 7.2.3. Recovering from Abnormal Closure in RFC 6544.
-                if backoff_delay == BACKOFF_MIN:
-                    initial_delay = random.random() * BACKOFF_INITIAL
-                    logger.info(
-                        "! connect failed; reconnecting in %.1f seconds",
-                        initial_delay,
-                        exc_info=True,
-                    )
-                    await asyncio.sleep(initial_delay)
-                else:
-                    logger.info(
-                        "! connect failed again; retrying in %d seconds",
-                        int(backoff_delay),
-                        exc_info=True,
-                    )
-                    await asyncio.sleep(int(backoff_delay))
-                # Increase delay with truncated exponential backoff.
-                backoff_delay = backoff_delay * BACKOFF_FACTOR
-                backoff_delay = min(backoff_delay, BACKOFF_MAX)
-                continue
-        else:
-            # Connection succeeded - reset backoff delay
-            backoff_delay = BACKOFF_MIN
-            refresh = True
+        except websockets.exceptions.InvalidStatusCode as e:
+            if refresh and e.status_code == 403:
+                new_token = await renew_token()
+                extra_headers["Authorization"] = f"Bearer {new_token}"
+                # Only attempt to refresh token once. If a new token cannot
+                # establish the connection, something else must cause 403
+                refresh = False
+            else:
+                raise  # abort
+        except websockets.exceptions.InvalidStatus as e:
+            if refresh and e.response.status_code == 403:
+                new_token = await renew_token()
+                extra_headers["Authorization"] = f"Bearer {new_token}"
+                # Only attempt to refresh token once. If a new token cannot
+                # establish the connection, something else must cause 403
+                refresh = False
+            else:
+                raise  # abort
+        except OSError as e:
+            if "[Errno 61]" in str(e):
+                # if connection cannot be established, retry later
+                backoff_delay = await _wait_before_retry(backoff_delay)
+            else:
+                raise  # abort
+        except websockets.exceptions.ConnectionClosedError as e:
+            if e.code == 1011:
+                # unexpected error raised from server
+                raise  # abort
+            else:
+                backoff_delay = await _wait_before_retry(backoff_delay)
+        except websockets.exceptions.ConnectionClosedOK:
+            backoff_delay = await _wait_before_retry(backoff_delay)
```

Comment:
@bzwei Shouldn't this be only called if there handler wants to have a retry.

Reply:
There is no problem to retry for both getting workload or sending event log. I removed the retry parameter. It is unlikely to meet connection close situation while getting the workload, but even if it happens we should be able to retry automatically.

Reply:
@bzwei Since the server controls when it is going to close the connection, if it closes the connection after it has given the worker payload will this cause a refetch of the payload.

Reply:
For reference this what the old code did when the server closed the connection
https://github.com/ansible/ansible-rulebook/blob/eac79808254997b47859d6422fd4fb4a0ffdaf2e/ansible_rulebook/websocket.py#L87